src/playermat/Playmat.ttslua

```diff
@@ -93,293 +70,271 @@ function onload(save_state)
+  elseif forcedLearning then
     printToColor("Drawing 2 cards, discard 1 (Forced Learning)", messageColor)
     drawCardsWithReshuffle(2)
+  else
+    drawCardsWithReshuffle(1)
+  end
 end
 
-function checkDeckZoneExists()
-    if getObjectFromGUID(zoneID) ~= nil then return end
-    spawnDeckZone()
+-- legacy function for "add draw 1 button"
+function doDrawOne(_, color)
+  setMessageColor(color)
+  drawCardsWithReshuffle(1)
 end
 
-function spawnDeckZone()
-    local pos = self.positionToWorld(DECK_POSITION)
-    local zoneProps = {
-      position = pos,
-      scale = DECK_ZONE_SCALE,
-      type = 'ScriptingTrigger',
-      callback = 'zoneCallback',
-      callback_owner = self,
-      rotation = self.getRotation()
-    }
-    spawnObject(zoneProps)
-end
+-- draw X cards (shuffle discards if necessary)
+function drawCardsWithReshuffle(numCards)
+  if type(numCards) ~= "number" then numCards = 1 end
+  getDrawDiscardDecks()
+
+  -- Norman Withers handling
+  if string.match(activeInvestigatorId, "%d%d%d%d%d") == "08004" then
+    local harbinger = false
+    if topCard ~= nil and topCard.getName() == "The Harbinger" then harbinger = true
+    elseif drawDeck ~= nil and not drawDeck.is_face_down then
+      local cards = drawDeck.getObjects()
+      if cards[#cards].name == "The Harbinger" then harbinger = true end
+    end
+
+    if harbinger then
+      printToColor("The Harbinger is on top of your deck, not drawing cards", messageColor)
+      return
+    end
+
+    if topCard ~= nil then
+      topCard.deal(numCards, PLAYER_COLOR)
+      numCards = numCards - 1
+      if numCards == 0 then return end
+    end
+  end
+
+  local deckSize = 1
+  if drawDeck == nil then
+    deckSize = 0
+  elseif drawDeck.tag == "Deck" then
+    deckSize = #drawDeck.getObjects()
+  end
 
-function zoneCallback(zone)
-    zoneID = zone.getGUID()
+  if deckSize >= numCards then
+    drawCards(numCards)
+    return
+  end
+
+  drawCards(deckSize)
+  if discardPile ~= nil then
+    shuffleDiscardIntoDeck()
+    Wait.time(|| drawCards(numCards - deckSize), 1)
+  end
+  printToColor("Take 1 horror (drawing card from empty deck)", messageColor)
 end
 
-function findObjectsAtPosition(localPos)
-    local globalPos = self.positionToWorld(localPos)
-    local objList = Physics.cast({
-        origin=globalPos, --Where the cast takes place
-        direction={0,1,0}, --Which direction it moves (up is shown)
-        type=2, --Type. 2 is "sphere"
-        size={2,2,2}, --How large that sphere is
-        max_distance=1, --How far it moves. Just a little bit
-        debug=false --If it displays the sphere when casting.
-    })
-    local decksAndCards = {}
-        for _, obj in ipairs(objList) do
-        if obj.hit_object.tag == "Deck" or obj.hit_object.tag == "Card" then
-            table.insert(decksAndCards, obj.hit_object)
-        end
+-- get the draw deck and discard pile objects
+function getDrawDiscardDecks()
+  drawDeck = nil
```

Comment:
Declare these at the top, or better yet make them local and return a table {drawDeck = ..., discardDeck = ...} (assuming you can put object refs in a table, I don't think I've ever tried that.

Reply:
will leave this as is for now since I don't wanna risk breaking anything that calls this